crates/payout_curve/src/lib.rs

```diff
@@ -0,0 +1,685 @@
+            if upper_event_outcome_payout > total_collateral {
+                upper_event_outcome_payout = total_collateral;
+            }
+            Ok((
+                PayoutPoint {
+                    event_outcome: lower_event_outcome.to_u64().expect("to fit into u64"),
+                    outcome_payout: lower_event_outcome_payout,
+                    extra_precision: 0,
+                },
+                PayoutPoint {
+                    event_outcome: upper_event_outcome,
+                    outcome_payout: upper_event_outcome_payout,
+                    extra_precision: 0,
+                },
+            ))
+        })
+        .collect::<Result<Vec<(_, _)>>>()?;
+
+    Ok(pieces)
+}
+
+/// Calculates the lower range payout points between 0 and the lower liquidation point
+fn calculate_lower_range_payouts(
+    coordinator_direction: Direction,
+    total_collateral: u64,
+    liquidation_price_lower_bound: Decimal,
+) -> Result<(PayoutPoint, PayoutPoint)> {
+    let (lower, upper) = match coordinator_direction {
+        // if coordinator is short, he gets everything from 0 until the trader's liquidation point
+        Direction::Short => (
+            PayoutPoint {
+                event_outcome: 0,
+                outcome_payout: total_collateral,
+                extra_precision: 0,
+            },
+            PayoutPoint {
+                event_outcome: liquidation_price_lower_bound
+                    .to_u64()
+                    .expect("to be able to fit decimal into u64"),
+                outcome_payout: total_collateral,
+                extra_precision: 0,
+            },
+        ),
+        // if coordinator is long, he gets 0 from 0 until his liquidation point
+        Direction::Long => (
+            PayoutPoint {
+                event_outcome: 0,
+                outcome_payout: 0,
+                extra_precision: 0,
+            },
+            PayoutPoint {
+                event_outcome: liquidation_price_lower_bound
+                    .to_u64()
+                    .expect("to be able to fit decimal into u64"),
+                outcome_payout: 0,
+                extra_precision: 0,
+            },
+        ),
+    };
+
+    Ok((lower, upper))
+}
+
+/// Calculates the upper range payout points between upper liquidation point and BTCUSD_MAX_PRICE
+fn calculate_upper_range_payouts(
+    coordinator_direction: Direction,
+    total_collateral: u64,
+    last_payout_point: PayoutPoint,
+) -> Result<(PayoutPoint, PayoutPoint)> {
+    let (lower_range_lower, lower_range_upper) = match coordinator_direction {
+        // if coordinator is long, he gets everything from the trader's liquidation point to
+        // infinity
+        Direction::Long => (
+            PayoutPoint {
+                event_outcome: last_payout_point.event_outcome,
+                outcome_payout: last_payout_point.outcome_payout,
```

Comment:
❓ Why is this not `total_collateral`?

Reply:
Unfortunately I found that sometimes we get a rounding error of 1 sat in the mid range's pnl. To ensure that the payout curve is continuity we just take the last point of the mid range and go until total collateral.